### PR TITLE
Add inference code

### PR DESCRIPTION
**Pull request by 0x434D**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,97 @@
-# TIRControlNet
\ No newline at end of file
+# TIR ControlNet
+Christian Mayr, Christian Kübler, Norbert Haala, Michael Teutsch,**"Diffusion based Thermal Semantic Segmentation"**, Perception Beyond the Visual Spectrum Workshop **PBVS**, Computer Vision and Pattern Recognition Conference **CVPR** - 2024, **Poster**
+
+
```
Comment (MichaelTeutsch) [inline README.md:4]: Please add the following line:
"Here is the direct link to the paper: [click me](https://openaccess.thecvf.com/content/CVPR2024W/PBVS/html/Mayr_Narrowing_the_Synthetic-to-Real_Gap_for_Thermal_Infrared_Semantic_Image_Segmentation_CVPRW_2024_paper.html)"

File: README.md [hunk 2]
```diff
@@ -1 +1,97 @@
-# TIRControlNet
\ No newline at end of file
+# TIR ControlNet
+Christian Mayr, Christian Kübler, Norbert Haala, Michael Teutsch,**"Diffusion based Thermal Semantic Segmentation"**, Perception Beyond the Visual Spectrum Workshop **PBVS**, Computer Vision and Pattern Recognition Conference **CVPR** - 2024, **Poster**
```
Comment (MichaelTeutsch) [inline README.md]: The title is
"Narrowing the Synthetic-to-Real Gap for Thermal Infrared Semantic Image Segmentation Using Diffusion-based Conditional Image Synthesis"
and not 
"Diffusion based Thermal Semantic Segmentation"

File: README.md [hunk 3]
```diff
@@ -1 +1,97 @@
...
+
+
+
+### Image Synthesis Preview
+
+The in the paper presented described approach for IR image synthesis based on provided semantics maps is demonstrated here.
```
Comment (MichaelTeutsch) [inline README.md]: The proposed approach for Thermal Infrared (TIR) image synthesis as described in the paper based on provided semantics maps is demonstrated here.

File: README.md [hunk 4]
```diff
@@ -1 +1,97 @@
...
+
+---
+
+![preview](assets/demo.png)
+ 
+---
+
+## Set Up on Your Own Machine
+
+To allow the research community to further experiment with ControlNet's ability to synthesize high quality IR data, we have decided to release our pretrained weights and inference script. This allows you to synthesize your own IR images.
```
Comment (MichaelTeutsch) [inline README.md]: ..ControlNet's ability synthesizing high-quality TIR imagery, we provide our pretrained...

Comment (MichaelTeutsch) [inline README.md]: Please replace all occurrences of "IR" with "TIR" to be consistent with the paper.

File: README.md [hunk 5]
```diff
@@ -1 +1,97 @@
...
+
+#### General
+A cuda capable GPU is required to run our code. There you should make sure to have to have your GPU setup correctly.
```
Comment (MichaelTeutsch) [inline README.md]: CUDA (all capital letters)

File: README.md [hunk 6]
```diff
@@ -1 +1,97 @@
...
+nvcc -V
+```
+```shell
+nvidia-smi
+```
+If there is no output you need to setup your GPU correctly. 
+
+#### Virtual Environment
+
+We highly recommend for you to use Conda as a package manager.
```
Comment (MichaelTeutsch) [inline README.md]: Remove "for you"

File: README.md [hunk 7]
```diff
@@ -1 +1,97 @@
...
+
+```shell
+# create virtual environment
+conda create -n IR_ConNet python=3.8
+conda activate IR_ConNet
+conda install pytorch torchvision torchaudio pytorch-cuda=12.1 -c pytorch -c nvidia
+conda install -c conda-forge diffusers transformers accelerate
+```
+
+#### Pretrained weights
```
Comment (MichaelTeutsch) [inline README.md]: Pretrained Weights (capital W)

File: README.md [hunk 8]
```diff
@@ -1 +1,97 @@
...
+
+The pretrained weights can be downloaded from HuggingFace.
+```shell
+https://huggingface.co/0x434D/TIR_ControlNet
+```
+
+#### Demo images
```
Comment (MichaelTeutsch) [inline README.md]: Demo Images (capital I)

File: README.md [hunk 9]
```diff
@@ -1 +1,97 @@
...
+
```
Comment (MichaelTeutsch) [inline README.md:56]: The demo images seem to be missing.

File: README.md [hunk 10]
```diff
@@ -1 +1,97 @@
...
+After setting up your virtual environment with conda and downloading the pretrained weights, you can simply use the provided inference script to run the sythesis using a given semantic map.
+
+
+## Important Notes
+
+### Grayscale
+The pretrained network was trained on the grayscale semantic maps. That is, the images that directly encode the label IDs. It is critical to use this kind of semantic maps and **not** colored ones.
+
+### Label ID convention - [FMB](https://arxiv.org/abs/2308.02097) Dataset
+The provided pretrained ControlNet has been trained using the [FMB](https://arxiv.org/abs/2308.02097) dataset and and is therefore trained on the conventions contained therein for the definition of label ids. If semantic maps originating from other datasets are supposed to be processed with the provided pretrained model the IDs have to be matched to the ones in the [FMB](https://arxiv.org/abs/2308.02097) dataset.
```
Comment (MichaelTeutsch) [inline README.md]: ...pretrained TIR ControlNet was trained...

File: README.md [hunk 11]
```diff
@@ -1 +1,97 @@
...
+| Sky           |  7 | (179,  73,  79)|
+| Person        |  8 | ( 76,  99, 166)|
+| Car           |  9 | ( 66, 121, 253)|
+| Truck         | 10 | (  6,   6,   6)|
+| Bus           | 11 | ( 12,  12,  12)|
+| Motorcycle    | 12 | (105, 153, 140)|
+| Bicycle       | 13 | (222, 215, 158)|
+| Pole          | 14 | (135, 113,  90)|
+
+### Size
```
Comment (MichaelTeutsch) [inline README.md]: Image Size

File: README.md [hunk 12]
```diff
@@ -1 +1,97 @@
...
+
+Note that the pretrained network is trained using 512 by 512 images - it is advised to stick to this size to achieve the best results.
```
Comment (MichaelTeutsch) [inline README.md]: ...using images with 512 x 512 pixels each - ...

File: README.md [hunk 13]
```diff
@@ -1 +1,97 @@
...
+
+
+
+## Citation
+
+If this work has been helpful to you, please feel free to cite our paper!
+
+```shell
+
```
Comment (MichaelTeutsch) [inline README.md]: @InProceedings{Mayr_2024_CVPR,
    author    = {Mayr, Christian and Kuebler, Christian and Haala, Norbert and Teutsch, Michael},
    title     = {{Narrowing the Synthetic-to-Real Gap for Thermal Infrared Semantic Image Segmentation Using Diffusion-based Conditional Image Synthesis}},
    booktitle = {Proceedings of the IEEE/CVF Conference on Computer Vision and Pattern Recognition (CVPR) Workshops},
    month     = {June},
    year      = {2024},
    pages     = {3131--3141}
}

File: inference.py [hunk 14]
```diff
@@ -0,0 +1,156 @@
+"""
```
Comment (MichaelTeutsch) [inline inference.py:22]: Please add license text at the top of the file. See here:
https://github.com/HensoldtOptronicsCV/FineGrainedVesselRecognition/blob/main/download.py

File: inference.py [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+        help="If true a random seed will be used, if False the seed is set to 1337",
+    )
+    # Parse the command-line arguments
+    arguments = parser.parse_args()
+
+    return arguments
+
+
+def pipeline_setup(base_model_path, contronlet_path):
+    """Controlnet pipeline setup
```
Comment (MichaelTeutsch) [inline inference.py]: ControlNet (capital N)

File: inference.py [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+        default=True,
+        help="If true a random seed will be used, if False the seed is set to 1337",
+    )
+    # Parse the command-line arguments
+    arguments = parser.parse_args()
+
+    return arguments
+
+
+def pipeline_setup(base_model_path, contronlet_path):
```
Comment (MichaelTeutsch) [inline inference.py]: controlnet_path

File: inference.py [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+    """Controlnet pipeline setup
+    
+        Input:
+            base_model_path - Path or link to the StableDiffusion base model.
+            controlnet_path - Path to the pretrained ControlNet for IR Image Synthesis
+        Output:
+            pipe - ControlNet pipeline object for image synthesis
+    """
+    controlnet = ControlNetModel.from_pretrained(
+        contronlet_path,
```
Comment (MichaelTeutsch) [inline inference.py]: controlnet_path

File: inference.py [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+
+import torch
+from diffusers import ControlNetModel, StableDiffusionControlNetPipeline
+from diffusers.utils import load_image
+
+
+def parse_arguments():
+    """Define argument parser"""
+    parser = argparse.ArgumentParser(description="Inference script")
+    # Adding positional argument
```
Comment (MichaelTeutsch) [inline inference.py]: Please better document arguments. Some of them are not fully self-explaining.

File: README.md (hunk 9 above)
Comment (0x434D) [reply README.md:56]: Demo images are located under the demo subfolder and work out of the box with the python script - should I mention this here ?

Comment (MichaelTeutsch) [reply README.md:56]: Probably yes. As a reader I would expect some text or images here after the heading "Demo Images".

File: inference.py [hunk 19]
```diff
@@ -1,7 +1,31 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+# Inspired by https://github.com/avaapm/marveldataset2016/blob/master/MARVEL_Download.py
```
Comment (MichaelTeutsch) [inline inference.py]: Please remove this line. This was only for the code of Heiko's paper.

File: inference.py [hunk 20]
```diff
@@ -19,66 +43,75 @@
...
+    # Prompt to be used by ControlNet's synthesis backbone
     parser.add_argument(
         "-p",
         "--prompt",
         type=str,
         default="Urban automotive scene containing vegetation, vehicles, roads, buildings, and persons",
         help="Prompt for Stable Diffusion",
     )
+
+    # The amount of inference steps ControlNet's synthesis backbone performs to synthesize an image
```
Comment (MichaelTeutsch) [inline inference.py]: The number of inference steps...

File: inference.py [hunk 21]
```diff
@@ -19,66 +43,75 @@
...
+    # Folder in which the synthesized images will be saved to
     parser.add_argument(
         "-o",
         "--output_folder",
         type=Path,
         default=(Path(__file__).resolve().parent / "results"),
         help="Default output folder - if set to None images won\'t be saved",
     )
+
+    # Flag wether to use a random or a set seed for image synthesis
```
Comment (MichaelTeutsch) [inline inference.py]: whether
